### PR TITLE
feat(flags): add timestamp-based flag reconstruction

### PR DESCRIPTION
**Pull request by gustavohstrassburger**
## Problem

Historical feature flag evaluation requires reconstructing flag definitions as they existed at specific points in time. Currently, we can only reconstruct flags by version number, but integrating with timestamps would enable more flexible historical evaluation workflows.

Part of #53999.

## Changes

- Add `find_version_at_timestamp` function to determine flag version at specific timestamps
- Add `reconstruct_flag_at_timestamp` convenience wrapper for timestamp-based reconstruction  
- Comprehensive test coverage for all timestamp-based reconstruction scenarios
- Handle edge cases like flags that didn't exist at given timestamps

## How did you test this code?

- Added comprehensive unit tests covering all scenarios: before creation, at creation, after updates, with missing versions
- All existing version history tests continue to pass
- Tests validate proper error handling for non-existent flags at timestamps

## LLM context

This PR implements the foundation for historical flag evaluation by adding timestamp-based flag reconstruction. The core functionality is isolated to the version history module and can be tested independently against the activity log without requiring UI components or Rust service integration.

The implementation efficiently queries the activity log using database filtering and stops at the first matching entry, avoiding unnecessary data scanning.

### CODE REVIEW

File: posthog/models/feature_flag/version_history.py [hunk 1]
```diff
@@ -168,3 +176,112 @@ def reconstruct_flag_at_version(
...
+    # Process entries in reverse order (oldest first) to track state transitions
+    entries_list = list(entries)
+    for activity, detail in reversed(entries_list):
+        # Extract version information from this entry first
+        changes = (detail or {}).get("changes") or []
+        version_after = _get_version_after(changes)
+
+        # Only consider entries that have version information
+        # Bulk delete entries without version changes should be skipped
+        if version_after is not None:
```
Comment (haacked) [inline posthog/models/feature_flag/version_history.py:243]: `blocking`: This skips state-tracking for any entry without a version change, but the comment above ("Bulk delete entries without version changes should be skipped") conflates two things: skipping the version read (fine) and skipping the deletion state (not fine). If a flag is bulk-deleted with `activity="deleted"` and empty `changes=[]` (the pattern your sibling test references from `api/feature_flag.py`), we never set `flag_was_deleted=True`, fall through to the `timestamp >= flag.updated_at` fallback, and return `flag.version` for a deleted flag.

Track the deletion state independently of version presence:

    if version_after is not None:
        most_recent_version = version_after

    # Track deletion/restoration regardless of whether this entry carried a
    # version change. Bulk deletes are logged with empty changes.
    if activity == "deleted":
        flag_was_deleted = True
    elif activity == "restored":
        flag_was_deleted = False

Worth adding a regression test: log `activity="deleted"` with `changes=[]`, then assert `find_version_at_timestamp` returns `None` at a later timestamp.

File: posthog/models/feature_flag/version_history.py [hunk 2]
```diff
@@ -168,3 +176,112 @@ def reconstruct_flag_at_version(
...
+
+    # We need to check the activity log to see if there were any deletions/restorations
+    # that happened between the flag's last update and the query timestamp
+    # We can't just return the current version when timestamp >= updated_at because
+    # there might be soft-delete operations recorded in the activity log after updated_at
+
+    # Get activity log entries at or before the timestamp, streaming with .iterator()
+    # to avoid loading all entries into memory. We need to check both deletion state
+    # and find the most recent entry that contains version information.
+    entries = (
```
Comment (haacked) [inline posthog/models/feature_flag/version_history.py:215]: `suggestion`: `.iterator()` is negated by `entries_list = list(entries)` a few lines down, so the streaming intent is lost. Also the three surrounding comments contradict each other ("Most recent first" on order_by, "Process entries in chronological order" on the state trackers, "Process entries in reverse order (oldest first)" on the loop).

The whole block reads cleaner as a newest-first early-exit, which also removes the state flags, the `reversed(list(...))`, and the ambiguity:

    entries = (
        ActivityLog.objects.filter(
            team_id=team_id,
            scope="FeatureFlag",
            item_id=str(flag.id),
            activity__in=["updated", "deleted", "restored"],
            created_at__lte=timestamp,
        )
        .order_by("-created_at", "-id")
        .values_list("activity", "detail")[:MAX_HISTORY_ENTRIES]
        .iterator()
    )

    for activity, detail in entries:
        changes = (detail or {}).get("changes") or []
        version_after = _get_version_after(changes)
        if activity == "deleted":
            return None
        if version_after is not None:
            return version_after

    if timestamp >= (flag.updated_at or flag.created_at):
        return flag.version
    return 1

(Also drop `flat=False`, it's the default for multi-column `values_list` and ignored.)

File: posthog/models/feature_flag/version_history.py [hunk 3]
```diff
@@ -168,3 +176,112 @@ def reconstruct_flag_at_version(
...
+
+        # Only consider entries that have version information
+        # Bulk delete entries without version changes should be skipped
+        if version_after is not None:
+            most_recent_version = version_after
+
+            # Track the deletion/restoration state transitions only for versioned entries
+            if activity == "deleted":
+                flag_was_deleted = True
+            elif activity == "restored":
```
Comment (haacked) [inline posthog/models/feature_flag/version_history.py:249]: `question`: `flag_was_deleted` is only cleared by `activity="restored"`. If an `updated` entry is logged after `deleted` without an intervening `restored` (data backfills, manual admin fixes, or a restore bug), the function returns `None` even though a later update implies the flag is live. Is that intentional? If so, worth a docstring note; if not, consider treating any versioned `updated` entry as implicit clearance of the deletion state.

File: posthog/models/feature_flag/version_history.py [hunk 4]
```diff
@@ -168,3 +176,112 @@ def reconstruct_flag_at_version(
...
+        return None
+
+    # If we found version information and flag wasn't deleted, return the version
+    if most_recent_version is not None:
+        return most_recent_version
+
+    # If we found no version changes but timestamp is after the flag's last update,
+    # return the current version (unless flag was deleted)
+    if timestamp >= (flag.updated_at or flag.created_at):
+        return flag.version
```
Comment (haacked) [inline posthog/models/feature_flag/version_history.py:263]: `suggestion`: This fallback returns `flag.version` when the scan finds no versioned entries. Combined with the bulk-delete skip above, that's how we leak a version for a deleted flag. Once the deletion-state tracking fix lands, this becomes safe, but worth a short comment here calling out the assumption (the scan has authoritative deletion state, or we wouldn't reach here).

File: posthog/models/feature_flag/version_history.py [hunk 5]
```diff
@@ -168,3 +176,112 @@ def reconstruct_flag_at_version(
...
+    # If we found no version changes, the flag was at version 1 (creation)
+    return 1
+
+
+def reconstruct_flag_at_timestamp(
+    flag: FeatureFlag,
+    timestamp: datetime,
+    team_id: int,
+) -> dict[str, Any]:
+    """
```
Comment (haacked) [inline posthog/models/feature_flag/version_history.py:274]: `suggestion`: The docstring still claims the wrapper exists "to avoid duplicating orchestration and reconstruction logic," but the wrapper runs two separate `ActivityLog` scans back-to-back (one in `find_version_at_timestamp`, one in `reconstruct_flag_at_version`). Either describe what the function does, or refactor to share the scan via a private helper (probably a follow-up once there's a real caller). Suggested wording:

    """
    Resolve the version active at `timestamp` and return that version's state.

    Raises VersionNotFound if the flag did not exist at the timestamp (before creation,
    or soft-deleted without a subsequent restore).
    Raises VersionHistoryIncomplete if the flag is missing version metadata, or if the
    activity log is missing entries needed for reconstruction.
    Raises ValueError if `timestamp` is not timezone-aware.
    """

File: posthog/models/feature_flag/test/test_version_history.py [hunk 6]
```diff
@@ -270,3 +274,255 @@ def test_reconstruct_field_cleared_to_none(self):
...
+    @parameterized.expand(
+        [
+            ("current_version", timedelta(hours=1), 3, 3),
+            ("before_creation", timedelta(hours=-1), 1, None),
+            ("at_creation", timedelta(0), 1, 1),
+        ]
+    )
+    def test_find_version_at_timestamp_boundary(self, _name, offset, flag_version, expected):
+        flag = self._create_flag(version=flag_version)
+        timestamp = flag.created_at + offset if _name != "current_version" else datetime.now(UTC)
```
Comment (haacked) [inline posthog/models/feature_flag/test/test_version_history.py:289]: `suggestion`: The `current_version` case uses `datetime.now(UTC)` while `flag.created_at` comes from a DB round-trip. Microsecond-precision differences between Python and Postgres can cause rare flakes. Using an offset from `flag.created_at` is deterministic, and lets you drop the `if _name != "current_version" else datetime.now(UTC)` special case:

```suggestion
        timestamp = flag.created_at + offset
```

Same pattern appears in `test_reconstruct_flag_at_timestamp_current_version`, `test_find_version_at_timestamp_soft_deleted_flag`, `_soft_deleted_then_restored`, and `_reconstruct_flag_at_timestamp_soft_deleted_flag`. Swap each `datetime.now(UTC)` for `flag.created_at + timedelta(hours=1)` (or `flag.updated_at + timedelta(hours=1)` after a delete log).

File: posthog/models/feature_flag/test/test_version_history.py [hunk 7]
```diff
@@ -270,3 +274,255 @@ def test_reconstruct_field_cleared_to_none(self):
...
+        log_activity(
+            organization_id=self.organization.id,
+            team_id=self.team.id,
+            user=self.user,
+            was_impersonated=False,
+            item_id=flag.id,
+            scope="FeatureFlag",
+            activity="deleted",
+            detail=Detail(
+                changes=[Change(type="FeatureFlag", action="changed", field="version", before=1, after=2)],
```
Comment (haacked) [inline posthog/models/feature_flag/test/test_version_history.py:451]: `suggestion`: This test logs a `deleted` entry with a version change (`field="version", before=1, after=2`), which matches the normal soft-delete path. The bulk-delete path (which the comment above references at `api/feature_flag.py`) logs `activity="deleted"` with `changes=[]`. That path is untested and is where the current code can incorrectly return `flag.version` for a deleted flag. Worth adding:

    def test_find_version_at_timestamp_bulk_deleted_flag(self):
        """Bulk deletes log activity=deleted with empty changes. Must still count as deleted."""
        flag = self._create_flag(name="V1", version=1)
        log_activity(
            organization_id=self.organization.id,
            team_id=self.team.id,
            user=self.user,
            was_impersonated=False,
            item_id=flag.id,
            scope="FeatureFlag",
            activity="deleted",
            detail=Detail(changes=[], name=flag.key),
        )
        future_timestamp = flag.created_at + timedelta(hours=1)
        assert find_version_at_timestamp(flag, future_timestamp, self.team.id) is None

File: posthog/models/feature_flag/test/test_version_history.py [hunk 8]
```diff
@@ -270,3 +274,255 @@ def test_reconstruct_field_cleared_to_none(self):
...
+        """Test version lookup between multiple updates to exercise activity log scanning."""
+        from posthog.models.activity_logging.activity_log import ActivityLog
+
+        flag = self._create_flag(name="V1", version=1)
+
+        # Simulate first update
+        self._simulate_update(flag, {"name": ("V1", "V2")})
+        # Fix timestamp for version 2 entries
+        first_update_time = flag.created_at + timedelta(minutes=10)
+        ActivityLog.objects.filter(
```
Comment (haacked) [inline posthog/models/feature_flag/test/test_version_history.py:316]: `suggestion`: This three-line `ActivityLog.objects.filter(... detail__changes__contains=...).update(created_at=...)` pattern repeats five times across the new tests. The `__contains` JSONB lookup silently matches on a subset, so if `_simulate_update` ever adds fields to `Change` this filter keeps matching but intent drifts. Worth extracting on the base class and asserting the update count:

    def _set_activity_timestamp(self, flag, version_after: int, when) -> None:
        updated = ActivityLog.objects.filter(
            team_id=self.team.id,
            scope="FeatureFlag",
            item_id=str(flag.id),
            detail__changes__contains=[{"field": "version", "after": version_after}],
        ).update(created_at=when)
        assert updated >= 1, f"no activity log entry found for version {version_after}"

Loud failure instead of silent mismatch.

File: posthog/models/feature_flag/version_history.py [hunk 9]
```diff
@@ -168,3 +176,112 @@ def reconstruct_flag_at_version(
...
+    entries = (
+        ActivityLog.objects.filter(
+            team_id=team_id,
+            scope="FeatureFlag",
+            item_id=str(flag.id),
+            activity__in=["updated", "deleted", "restored"],
+            created_at__lte=timestamp,  # Only entries at or before the timestamp
+        )
+        .order_by("-created_at", "-id")  # Most recent first
+        .values_list("activity", "detail", flat=False)[:MAX_HISTORY_ENTRIES]
```
Comment (haacked) [inline posthog/models/feature_flag/version_history.py:224]: `nit`: `flat=False` is the default for multi-column `values_list` and is ignored.

```suggestion
        .values_list("activity", "detail")[:MAX_HISTORY_ENTRIES]
```

**Review by haacked — COMMENTED** [review]
Thanks for splitting this out. There's one blocking issue around bulk deletes and a few other suggestions and nits. Other than that, everything looks good!

File: posthog/models/feature_flag/version_history.py [hunk 10]
```diff
@@ -168,3 +176,88 @@ def reconstruct_flag_at_version(
...
+        if version_after is not None:
+            return version_after
+
+    # No activity log entries found. If timestamp is after the flag's last update,
+    # return the current version. The scan has authoritative deletion state,
+    # or we wouldn't reach here.
+    if timestamp >= (flag.updated_at or flag.created_at):
+        return flag.version
+
+    return 1
```
Comment (haacked) [inline posthog/models/feature_flag/version_history.py]: `suggestion`: This `return 1` only runs when the scan found nothing at or before `timestamp` and `timestamp < flag.updated_at`. In the common case (flag defaults to `version=1` at creation) that's correct, but for a flag created with a non-default initial version, or one with a pruned activity log, this silently returns the wrong version. Since we can't see the `before` value of the first post-timestamp update from here, we can't recover the real answer.

Two options, your call:

1. Raise `VersionHistoryIncomplete` so corrupt/incomplete history surfaces instead of being papered over.
2. Keep `return 1` but pin the invariant: add a comment noting this assumes `flag.version == 1` at creation, and consider `assert flag.version == 1` (defensive) since this branch is reached rarely.

File: posthog/models/feature_flag/test/test_version_history.py [hunk 11]
```diff
@@ -270,3 +274,273 @@ def test_reconstruct_field_cleared_to_none(self):
...
+
+    def test_reconstruct_flag_at_timestamp_current_version(self):
+        flag = self._create_flag(name="Current", version=2)
+        future_timestamp = flag.created_at + timedelta(hours=1)
+
+        result = reconstruct_flag_at_timestamp(flag, future_timestamp, self.team.id)
+
+        assert result["name"] == "Current"
+        assert result["version"] == 2
+        assert result["is_historical"] is False
```
Comment (haacked) [inline posthog/models/feature_flag/test/test_version_history.py:376]: `suggestion`: This is the only test exercising the `is_historical=False` branch of `_build_response`, so worth pinning the other fields that branch populates. A regression that returned `None` for `version_timestamp` or swapped `modified_by` for `created_by_id` wouldn't be caught today.

```suggestion
        assert result["is_historical"] is False
        assert result["version_timestamp"] == (flag.updated_at or flag.created_at)
        assert result["modified_by"] == flag.last_modified_by_id
```

File: posthog/models/feature_flag/test/test_version_history.py [hunk 12]
```diff
@@ -270,3 +274,273 @@ def test_reconstruct_field_cleared_to_none(self):
...
+    def test_find_version_at_timestamp_naive_timestamp_raises(self):
+        flag = self._create_flag(version=1)
+        naive_timestamp = datetime(2024, 1, 1, 12, 0, 0)  # No timezone
+        with self.assertRaises(ValueError) as cm:
+            find_version_at_timestamp(flag, naive_timestamp, self.team.id)
+        assert "timezone-aware" in str(cm.exception)
+
+    def test_find_version_at_timestamp_with_updates(self):
+        """Test version lookup between multiple updates to exercise activity log scanning."""
+        from posthog.models.activity_logging.activity_log import ActivityLog
```
Comment (haacked) [inline posthog/models/feature_flag/test/test_version_history.py:308]: `nit`: This module is already imported at the top of the file for `log_activity`, `Change`, `ChangeAction`, `Detail`. Worth adding `ActivityLog` to that import line and dropping the inline ones here and in `test_find_version_at_timestamp_skips_entries_without_version`.

File: posthog/models/feature_flag/test/test_version_history.py [hunk 13]
```diff
@@ -270,3 +274,273 @@ def test_reconstruct_field_cleared_to_none(self):
...
+            detail=Detail(
+                changes=[Change(type="FeatureFlag", action="changed", field="version", before=2, after=3)],
+                name=flag.key,
+            ),
+        )
+
+        # Query after restoration should return the restored version
+        future_timestamp = flag.created_at + timedelta(hours=1)
+        version = find_version_at_timestamp(flag, future_timestamp, self.team.id)
+        assert version == 3
```
Comment (haacked) [inline posthog/models/feature_flag/test/test_version_history.py:498]: `nit`: The `deleted` and `restored` entries aren't backdated, so the test depends on wall-clock ordering between two consecutive `log_activity` calls (the `-created_at` sort picks `restored` because it's logged second). It works, but the temporal intent is incidental. A two-line `.filter(...).update(created_at=...)` for each entry makes it explicit and removes the dependency on ordering:

    deleted_time = flag.created_at + timedelta(minutes=5)
    restored_time = flag.created_at + timedelta(minutes=10)
    ActivityLog.objects.filter(
        team_id=self.team.id, scope="FeatureFlag", item_id=str(flag.id), activity="deleted"
    ).update(created_at=deleted_time)
    ActivityLog.objects.filter(
        team_id=self.team.id, scope="FeatureFlag", item_id=str(flag.id), activity="restored"
    ).update(created_at=restored_time)

(Related: the `_set_activity_timestamp` helper from the prior review would collapse this pattern across the file.)

**Review by haacked — APPROVED** [review]
Nice work! You've addressed all my previous feedback except for the `_set_activity_timestamp` helper suggestion from last round (feel free to disagree with whether its needed or not).

A few more minor nits and suggestions, but nothing blocking.

**Comment by gustavohstrassburger** [comment]
thank you for the thorough review @haacked,
I'll merge this as is, but I'll address these comments in one of the other PRs.
